src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs

```diff
@@ -324,11 +326,11 @@ public bool TryGetProperty(ReadOnlySpan<char> propertyName, out JsonElement valu
                 return document.TryGetNamedPropertyValue(_idx, propertyName, out value);
             }
 
-            var jsonNode = (JsonNode)_parent;
+            var jsonNode = (JsonNode)_parent!;
```

Comment:
`[DoesNotReturnIf (true)]` doesn't seem solve that as we need to evaluate/set the Boolean value before call. We could use `[NotNullIfNotNull("parent")] ` but that returns the value which need to consumed on caller side. Not sure if we want such change:
```csharp
private object CheckValidInstance([NotNullIfNotNull("parent")] object? parent)
{
    if (parent == null)
    {
         throw new InvalidOperationException();
     }

     Debug.Assert(_parent is JsonDocument || _parent is JsonNode);
     return parent;
}

internal bool TextEqualsHelper(ReadOnlySpan<char> text, bool isPropertyName)
{
    object parent = CheckValidInstance(_parent);

    var document = (JsonDocument)parent;
    return document.TextEquals(_idx, text, isPropertyName);
}

Reply:
> `[DoesNotReturnIf (true)]` doesn't seem solve that as we need to evaluate/set the Boolean value before call. 

Right, is doing that a concern?

The call site would be:
```C#
CheckValidInstance_ThrowInvalidOperationExceptionIfNull(_parent == null, _parent);
```

> We could use `[NotNullIfNotNull("parent")] ` but that returns the value which need to consumed on caller side. Not sure if we want such change

That might be OK too. That would mean that all uses of the `_parent` field would need to be replaced with a local object that `CheckValidInstance` returned. Although, doing that is more invasive than just using `!`.

Reply:
> Right, is doing that a concern?

Nope, i misread your example. I'm OK with it. 

Not sure If we end up to make it non null instead https://github.com/dotnet/runtime/pull/528#discussion_r355625970

Reply:
I'm not sure what `CheckValidInstance_ThrowInvalidOperationExceptionIfNull(_parent == null, _parent);` is trying to solve, but anything that involves repeating code over and over instead of letting the instance method do the right thing seems wrong in my book.

Reply:
It tries to solve using `!` to workaround compiler warnings. The instance method `CheckValidInstance` does the right thing, but the compiler still complains after the fact that `_parent` could be null. By making the check static and annotating that this method doesn't return if `_parent == null`, we help the compiler with code flow analysis.